synapse/federation/federation_server.py

```diff
@@ -1109,16 +1116,30 @@ async def _process_incoming_pdus_in_room_inner(
                         (self._clock.time_msec() - received_ts) / 1000
                     )
 
-            # We need to do this check outside the lock to avoid a race between
-            # a new event being inserted by another instance and it attempting
-            # to acquire the lock.
-            next = await self.store.get_next_staged_event_for_room(
-                room_id, room_version
-            )
-            if not next:
+            while True:
+                # We need to do this check outside the lock to avoid a race between
+                # a new event being inserted by another instance and it attempting
+                # to acquire the lock.
+                next = await self.store.get_next_staged_event_for_room(
+                    room_id, room_version
+                )
+
+                if next is None:
+                    break
+
+                origin, event = next
+
+                if await self._spam_checker.drop_federated_event(event):
+                    logger.warning(
+                        "Staged federated event contains spam, dropping %s",
+                        event.event_id,
+                    )
+                    continue
+
```

Comment:
I think its probably worth splitting this out into a helper function, as having nested loops is causing my brain to melt slightly.

Reply:
splitting out the whole nested while, i take it? like `_get_next_valid_staged_event_for_room` or something?

Reply:
Yup, exactly